### PR TITLE
Implemented Bn254 field

### PR DESCRIPTION
**Pull request by utsavjnn**
Lets review and merge this and we can add Bn254 hashing later when bandwidth allows.

@ultrainstinct30

### CODE REVIEW

File: field/src/bn254.rs [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+
+impl Field for Bn254Field {
+    const ZERO: Self = Self([0; 4]);
+
+    const ONE: Self = Self([1, 0, 0, 0]);
+
+    const TWO: Self = Self([2, 0, 0, 0]);
+
+    const NEG_ONE: Self = Self([
+        4891460686036598785 as u64,
```
Comment (ultrainstinct30) [inline field/src/bn254.rs]: The first limb will be `4891460686036598784`.

File: field/src/bn254.rs [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+    fn try_inverse(&self) -> Option<Self> {
+        if self.is_zero() {
+            return None;
+        }
+
+        // Fermat's Little Theorem
+        Some(self.exp_biguint(&(Self::order() - BigUint::one() - BigUint::one())))
+    }
+
+    fn from_noncanonical_biguint(n: BigUint) -> Self {
```
Comment (ultrainstinct30) [inline field/src/bn254.rs:123]: Taking a modulus here with the field order is necessary.

File: field/src/bn254.rs [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+use core::fmt::{Debug, Display, Formatter};
+use core::hash::{Hash, Hasher};
+use core::iter::{Product, Sum};
+use core::ops::{Add, AddAssign, Div, DivAssign, Mul, MulAssign, Neg, Sub, SubAssign};
+use itertools::Itertools;
+use serde::{Serialize, Deserialize};
+use num::bigint::BigUint;
+use num::{Integer, One};
+use rand::RngCore;
+use std::str::FromStr;
```
Comment (ultrainstinct30) [inline field/src/bn254.rs]: This import should be from `core` rather than `std` I think.

**Review by ultrainstinct30 — CHANGES_REQUESTED** [review]

File: field/src/bn254.rs (hunk 1 above)
Comment (utkarsh-21st) [reply field/src/bn254.rs]: done

File: field/src/bn254.rs (hunk 2 above)
Comment (utkarsh-21st) [reply field/src/bn254.rs:123]: done

File: field/src/bn254.rs (hunk 3 above)
Comment (utkarsh-21st) [reply field/src/bn254.rs]: done

**Pull request merged**